torch/csrc/jit/passes/restore_mutation.cpp

```diff
@@ -0,0 +1,136 @@
+#include <ATen/core/interned_strings.h>
+#include <ATen/core/jit_type.h>
+#include <torch/csrc/jit/passes/remove_mutation.h>
+#include <torch/csrc/jit/passes/restore_mutation.h>
+
+namespace torch {
+namespace jit {
+namespace {
+static const std::unordered_set<Symbol>
+    functional_activation_possible_type_promotion =
+        {aten::hardsigmoid, aten::hardswish, aten::sigmoid, aten::tanh};
+
+static const std::unordered_set<Symbol>
+    functional_activation_no_type_promotion = {
+        aten::hardtanh,
+        aten::relu,
+        aten::relu6};
+
+static const std::unordered_set<Symbol> inplace_activation = []() {
+  std::unordered_set<Symbol> target_ops;
+  for (auto op : functional_activation_possible_type_promotion) {
+    std::string name = std::string(op.toQualString()) + "_";
+    target_ops.insert(Symbol::fromQualString(name));
+  }
+  for (auto op : functional_activation_no_type_promotion) {
+    std::string name = std::string(op.toQualString()) + "_";
+    target_ops.insert(Symbol::fromQualString(name));
+  }
+  return target_ops;
+}();
+
+} // namespace
+
+FunctionalToInplaceRewriter::FunctionalToInplaceRewriter(
+    std::shared_ptr<Graph> graph,
+    c10::optional<std::function<FunctionalToInplaceRewriterFlags(Node*)>>
+        filter)
+    : aliasDb_(nullptr),
+      graph_(std::move(graph)),
+      node_filter_(std::move(filter)) {}
+
+bool FunctionalToInplaceRewriter::CanBeInplace(Node* node) {
+  bool check_shape = true;
+  bool check_dtype = true;
+
+  if (node_filter_) {
+    auto flags = (*node_filter_)(node);
+    if (!flags.transform) {
+      return false;
+    }
+    check_shape = flags.check_shape;
+    check_dtype = flags.check_dtype;
+  }
+
+  Value* input = node->inputs().at(0);
+  Value* output = node->output();
+  if (input->type() != TensorType::get() ||
+      output->type() != TensorType::get()) {
+    return false;
+  }
+  auto inputDtype = input->type()->cast<TensorType>()->scalarType();
+  auto outputDtype = output->type()->cast<TensorType>()->scalarType();
+
+  // if (check_shape) {
+  //  return false;
+  //}
+
+  // In general, we don't need to check shape as activation ops are
```

Comment:
We're not checking that there's an equivalent inplace schema for the node when making this transformation, which we know for activations but we don't know if we're making this a public exposed api

Reply:
I will add a schema checking.